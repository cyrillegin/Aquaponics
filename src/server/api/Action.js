import { Router } from 'express';

const router = new Router();

/**
 * POST
 * Creates a new action
 *
 * params: {
 *   condition: 'if value > 5',
 *   action: 'send slack',
 *   interval: '5d',
 *   stationId: 1,
 *   sensorId: 1
 * }
 *
 * returns: 200 success, 400 if validation fails
 */
router.post('/', async (req, res) => {
  console.info('POST request to action');
  const { stationId, sensorId, condition, action, interval } = req.body;

  const valid = validateActionParams(req.body);
  if (valid.error) {
    res.status(400).send({ error: valid.error });
    return;
  }

  try {
    const result = await Action.create({ stationId, sensorId, condition, action, interval });
    console.info('new action added');
    res.status(200).send({ message: 'success' });
  } catch (e) {
    console.error('an error occured!');
    console.error(e);
    res.status(400).send({ error: 'An unknown error has occured' });
  }
});

/**
 * DELETE
 * Deletes an existing action
 *
 * kwargs: id=stationId (int)
 *
 * returns: 200 - success
 *          400 if id isn't given
 */
router.delete('/', async (req, res) => {
  console.info('DELETE request to action');
  if (!req.query.id) {
    res
      .status(400)
      .send({ error: 'You must provide the id of the action you would like to delete' });
    return;
  }
  try {

    const actions = await Action.destroy({
      where: {id: req.query.id}
    })

    res.status(200).send({ message: `success`, actions });
  } catch (e) {
    console.error('an error occured!');
    console.error(e);
    res.status(400).send({ error: 'An unknown error has occured' });
  }
});

const validateActionParams = params => {
  if(!params.stationId) {
    return { error: 'Station id required };
  }
  if(!params.sensorId) {
    return { error: 'Sensor id required };
  }
  if(!params.condition) {
    return { error: 'Condition required };
  }
  if(!params.action) {
    return { error: 'Action required };
  }
  if(!params.interval) {
    return { error: 'Interval required };
  }
  return {}
}

export default router;
