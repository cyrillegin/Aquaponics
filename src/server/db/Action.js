const { Model, DataTypes } = require('sequelize');

class Action extends Model {}

function buildActionSchema(sequelize) {
  Action.init(
    {
      stationId: { type: DataTypes.INTEGER, allowNull: false, field: 'station_id' },
      sensorId: { type: DataTypes.INTEGER, allowNull: false, field: 'sensor_id' },
      condition: { type: DataTypes.STRING, allowNull: false },
      action: { type: DataTypes.STRING, allowNull: false },
      value: { type: DataTypes.STRING, allowNull: false },
      metadata: { type: DataTypes.STRING },
      interval: { type: DataTypes.STRING, allowNull: false },
      // Can be one of: 'value', 'timestamp', 'time'
      valueType: {
        type: DataTypes.STRING,
        allowNull: false,
        defaultValue: 'value',
        field: 'value_type',
      },
    },
    {
      sequelize,
      modelName: 'action',
      createdAt: 'created_at',
      updatedAt: 'updated_at',
      paranoid: true,
      underscored: true,
    },
  );
}

export { buildActionSchema, Action };
