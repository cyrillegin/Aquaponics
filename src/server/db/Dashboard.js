import { Model, DataTypes } from 'sequelize';

class Dashboard extends Model {}

function buildDashboardSchema(sequelize) {
  Dashboard.init(
    {
      name: { type: DataTypes.STRING, allowNull: false },
      dashboardId: { type: DataTypes.STRING, allowNull: false, field: 'dashboard_id' },
      sensorId: { type: DataTypes.INTEGER, allowNull: false, field: 'sensor_id' },
      station_Id: { type: DataTypes.INTEGER, allowNull: false, field: 'station_id' },
      position: { type: DataTypes.INTEGER, allowNull: false },
    },
    {
      sequelize,
      modelName: 'dashboard',
      createdAt: 'created_at',
      updatedAt: 'updated_at',
      paranoid: true,
      underscored: true,
    },
  );
}

export { buildDashboardSchema, Dashboard };
