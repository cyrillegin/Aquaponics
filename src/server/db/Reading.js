const { Model, DataTypes } = require('sequelize');

class Reading extends Model {}

function buildReadingSchema(sequelize) {
  Reading.init(
    {
      sensorId: { type: DataTypes.INTEGER, allowNull: false, field: 'sensor_id' },
      stationId: { type: DataTypes.INTEGER, allowNull: false, field: 'station_id' },
      value: { type: DataTypes.DOUBLE, allowNull: false },
      timestamp: { type: DataTypes.DATE, allowNull: false },
    },
    {
      sequelize,
      modelName: 'reading',
      createdAt: 'created_at',
      updatedAt: 'updated_at',
      paranoid: true,
      underscored: true,
    },
  );
}

export { buildReadingSchema, Reading };
