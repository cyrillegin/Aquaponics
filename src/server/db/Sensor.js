import { Model, DataTypes } from 'sequelize';
import SENSOR_TYPES from '../constants';

class Sensor extends Model {}

function buildSensorSchema(sequelize) {
  Sensor.init(
    {
      name: { type: DataTypes.STRING, allowNull: false },
      station_id: { type: DataTypes.UUID, allowNull: false },
      coefficients: { type: DataTypes.STRING, allowNull: true },
      dexcription: { type: DataTypes.STRING, allowNull: true },
      on: { type: DataTypes.BOOLEAN, allowNull: true },
      type: { type: DataTypes.ENUM, values: SENSOR_TYPES, allowNull: false },
    },
    { sequelize, modelName: 'sensor' },
  );
}

export { Sensor, buildSensorSchema };
