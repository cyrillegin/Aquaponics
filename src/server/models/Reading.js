import BaseModel from './BaseModel';
import ReadingMeta from './ReadingMeta';

export default class Reading extends BaseModel {
  validator = ReadingMeta.validator;
  indexes = ReadingMeta.indexes;

  constructor(context, collectionName) {
    super(context, 'reading');
    this.context = context;
    this.collection = context.db.collection('reading');
  }

  async getSensorByStation() {
    return await this.collection
      .aggregate([{ $group: { _id: { name: '$name', station: '$station' } } }])
      .toArray();
  }
}
