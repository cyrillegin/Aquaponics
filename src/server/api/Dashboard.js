import { Router } from 'express';
import { Dashboard, DashboardGraph, DashboardSource } from '../db';
import { validateDashboardParams } from '../utilities/Validators';

const router = new Router();

/**
 * GET
 * Gets an array of dashboards
 *
 * returns {
 *   dashboards: [{
 *     id: 123,
 *     name: 'asdf',
 *     dashboardGraph: [{
 *       dashboardId: 123,
 *       id: 123,
 *       width: 123,
 *       position: 123,
 *       type: 'asdf',
 *       title: 'asdf'
 *    }],
 *  }]
 * }
 */
router.get('/', async (req, res) => {
  console.debug('GET request to dashboard');
  const dashboards = await Dashboard.findAll({
    include: [
      {
        model: DashboardGraph,
        include: [DashboardSource],
      },
    ],
  });

  res.send(dashboards);
});

/**
 * POST
 * Creates a new dashboard
 *
 * params: {
 *   name: 'station name'
 * }
 *
 * returns: 200 success, 400 if validation fails
 */
router.post('/', async (req, res) => {
  console.debug('POST request to dashboard');
  const { name } = req.body;

  const valid = validateDashboardParams(req.body);
  if (valid.error) {
    res.status(400).send({ error: valid.error });
    return;
  }

  try {
    Dashboard.create({ name });
    console.debug('new dashboard created.');
    res.status(200).send({ message: 'success' });
  } catch (error) {
    console.error('an error occured!');
    console.error(error);
    res.status(400).send({ error: 'An unknown error has occured' });
  }
});

export default router;
