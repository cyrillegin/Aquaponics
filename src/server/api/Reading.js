import { Router } from 'express';
import { Op } from 'sequelize';
import { Reading } from '../db';
import { validateReadingParams } from '../utilities/Validators';

const router = new Router();

/**
 * GET
 * Gets an array of all the readings for a particular sensor
 *
 * kwargs: sensorId (int)
 *         startTime (date) - default 7 days ago
 *         endTime (date) - default now
 *
 * returns {
 *   readings: [{
 *     value: 123,
 *     timestamp: '10-4'
 *   }]
 * }
 */
router.get('/', async (req, res) => {
  console.info('GET request to reading');
  let { sensorId, start, end } = req.query;
  if (!sensorId) {
    res.status(400).send({ error: 'You must provide a sensor id' });
    return;
  }

  const currentDate = new Date();
  const tempDate = new Date();

  if (!start) {
    tempDate.setDate(currentDate.getDate() - 7);
    start = tempDate;
  }

  if (!end) {
    end = currentDate;
  }

  if (end < start) {
    tempDate.setDate(new Date(end).getDate() - 7);
    start = tempDate;
  }

  const readings = await Reading.findAll({
    where: {
      sensorId,
      timestamp: {
        [Op.lt]: end,
        [Op.gt]: start,
      },
    },
  });
  res.send(readings);
});

/**
 * POST
 * Creates a new station
 *
 * params: {
 *   name: 'station name',
 *   ip: 'xxx.xxx.xxx.xxx',
 *   sensorId: 1,
 *   stationId: 1
 * }
 *
 * returns: 200 success, 400 if validation fails
 */
router.post('/', async (req, res) => {
  console.info('POST request to reading');
  console.log(req.body)
  const { value, timestamp, sensorId, stationId } = req.body;


  const valid = validateReadingParams(req.body);
  if (valid.error) {
    res.status(400).send({ error: valid.error });
    return;
  }

  try {
    const result = await Reading.create({ value, timestamp, sensorId, stationId });
    console.info('new reading added');
    res.status(200).send({ message: 'success' });
  } catch (e) {
    console.error('an error occured!');
    console.error(e);
    res.status(400).send({ error: 'An unknown error has occured' });
  }
});

export default router;
