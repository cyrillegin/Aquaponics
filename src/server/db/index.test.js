import { buildStationSchema, Station } from './Station';
import { buildSensorSchema, Sensor } from './Sensor';
import { buildReadingSchema, Reading } from './Reading';
import { buildActionSchema, Action } from './Action';
import { buildDashboardSchema, Dashboard } from './Dashboard';

jest.mock('sequelize');

describe('database', () => {
  it('should take a snapshot of the schema', () => {
    const takeSnap = schema => {
      expect(Object.keys(schema)).toMatchSnapshot();
    };
    Station.init = takeSnap;
    Sensor.init = takeSnap;
    Reading.init = takeSnap;
    Action.init = takeSnap;
    Dashboard.init = takeSnap;
    buildStationSchema();
    buildSensorSchema();
    buildReadingSchema();
    buildActionSchema();
    buildDashboardSchema();
  });
});
