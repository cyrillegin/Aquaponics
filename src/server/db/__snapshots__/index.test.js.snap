// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`database should take a snapshot of the schema 1`] = `
Array [
  "name",
  "address",
  "port",
]
`;

exports[`database should take a snapshot of the schema 2`] = `
Array [
  "name",
  "stationId",
  "coefficients",
  "description",
  "hardwareName",
  "hardwareType",
  "readingType",
  "poll_rate",
  "unit",
]
`;

exports[`database should take a snapshot of the schema 3`] = `
Array [
  "sensorId",
  "stationId",
  "value",
  "timestamp",
]
`;

exports[`database should take a snapshot of the schema 4`] = `
Array [
  "stationId",
  "sensorId",
  "condition",
  "action",
  "value",
  "metaData",
  "interval",
  "valueType",
]
`;

exports[`database should take a snapshot of the schema 5`] = `
Array [
  "name",
  "dashboard_id",
  "sensor_id",
  "station_id",
  "position",
]
`;
